### ISSUE TITLE
Feature Request - Precipitation Type

### ISSUE THREAD

**Issue by Craytor** [tsupinie/autumnplot-gl#7]
This package has been amazing to work with! I'm wondering about the possibility of adding precipitation data to model data. Models have reflectivity, and they also have information about precip type. I'm wondering about the possibility of supporting precip type down the road. I know this is a rather complex task, but would we be able to achieve this using a "filter" in WebGL to apply a separate color palette based on the precipitation type? Finally, for the required data, in terms of the current project, this would require three or four items in addition to reflectivity - `CSNOW`, `CICEP`, `CFRZR`, and possibly `CRAIN`. Perhaps CRAIN may not be needed as if your precip type is not snow, mix, or freezing rain, it's going to be rain?

Thoughts on this? In terms of structure - perhaps this should be an additional layer type such as a "Shaded Layer" and perhaps the inputs could be data, and then the layer types. I think that this should also be applicable to both ContourFilled and Raster. 

I'm willing to help out any way possible, but my knowledge of WebGL is lacking.

**Comment by tsupinie** [comment]
Yeah, I think this could be an interesting addition. I'm trying to mull over in my head exactly how I would want it to look and how I would want the code to be structured. But it's doable, and making it part of the library would be more efficient than just having the user add four layers if they wanted a ptype display.

**Comment by Craytor** [comment]
I was curious if you've given this any more thought on a "best approach."

**Comment by tsupinie** [comment]
The short answer is "yes, but ..."

Basically, my plan was to allow the user to pass a mask to the ContourFill and Raster classes that would select from a bunch of different colormaps that the user could pass in as well. The problem with that is, due to the way I'm implementing colormaps right now, each colormap takes two textures. 

The maximum number of textures (per draw call) that's guaranteed to be supported on all hardware is 8, so a colormap taking two textures is a bit wasteful in that context. If a user wanted to do something like PivotalWeather's precip type/precip rate plot (attached), they'd need one texture for the precip rate, one for the precip type mask, and eight for the colormaps. That's 10 textures, so above that limit.

![prateptype_cat-imp conus](https://github.com/user-attachments/assets/19a4d5ab-9f57-4f64-8acf-8c8577246655)

So I went down a rabbit hole of trying to get each colormap down to one texture, which for *WebGL reasons* is gonna be trickier than I thought it was gonna be when I started. (Plus, I realized I might have a fencepost bug in colormaps for Raster plots, which I'll need to fix somehow.) I haven't worked on this in a bit, so I might pick it up again soon.

*gets up, walks around, makes lunch*

Though now that I think on it for a few minutes ...

I could still have that API and basically draw each color separately. So in the precip type/precip rate example, you'd have one draw call for the rain ptype, one for the snow ptype, etc., and it would be transparent to the user. That would get around that WebGL limitation, and I wouldn't be stuck on the colormap optimization.

Well, thanks for the unintentional [rubber-ducking](https://en.wikipedia.org/wiki/Rubber_duck_debugging)!

**Comment by tsupinie** [comment]
I've modified the `Raster` and `ContourFill` components to take multiple colormaps and a mask, which allows you to do a precip type plot in one call. It's called like this:

```javascript
const raster_cref = new Raster(raw_cref_field, {cmap: [crain_cmap, csnow_cmap, cicep_cmap, cfrzr_cmap], 
                                                cmap_mask: data_mask});
```

`data_mask` is a `Uint8Array` on the same grid as the main field. I struggled with whether to pass the mask as a bare array or as a `RawScalarField`, but the types in Typescript work out more intuitively (I think) if I just pass a `Uint8Array`.

<img width="1584" alt="Image" src="https://github.com/user-attachments/assets/8e2b7fce-cd71-4556-8188-7c8613a95568" />

I haven't pushed this quite yet. I'll need to do some cleanup and documentation, but once I push, I'll let you know.

**Comment by tsupinie** [comment]
Okay, this is pushed to the development branch. Let me know if that works for you, and it'll go in when I release the next version.

**Comment by Craytor** [comment]
Sorry, just saw this! This is beautiful, I will take a look tomorrow.

**Comment by tsupinie** [comment]
Oh, yeah, I guess I never said in here how to fill `cmap_mask`. A value of 1 uses the first colormap, a value of 2 uses the second colormap, etc. Values of 0 or any value more than the number of colormaps (so 5 or above if you have 4 colormaps) won't get plotted.

**Comment by Craytor** [comment]
Ok, that makes a lot of sense. This all seems perfect. Was it as bad as you were expecting to pull it off?

I started to play around with it. I did notice this occurring if you "superzoom" into a specific area if it's not using the first cmap. An example of this is in the DropBox video linked below.

https://www.dropbox.com/scl/fi/h5k0yhsq67jj44sgjlf2j/Screen-Recording-2025-02-17-at-3.54.40-PM.mov?rlkey=yqi2rxoghxqlsnme7dqsay47y&st=svgt757f&dl=0

**Comment by tsupinie** [comment]
Like I implied above, the part that tripped me up the most was trying to figure out how to make the types work. I wanted to pass the mask as a `RawScalarField`, but that would have required templating the `ContourFillOptions` for an optional argument, which didn't make a whole lot of sense. In the end I figured the least-bad option was to have the mask be passed as a UInt8Array. Otherwise, it ended up being pretty much what I expected except for your video, which is, uh, interesting. It has something to do with the web Mercator projection because it occurs at any zoom level if I turn off the globe view. (And also if I degrade maplibre to v3.0.) I'll have to fix that. Thanks for pointing that out!

**Comment by tsupinie** [comment]
Ohhhhhh, found it. I wasn't resetting one of the WebGL uniforms at the beginning of the colormap loop like I should have been. Should be fixed in the development branch now.

**Comment by Craytor** [comment]
Aha, it's working perfect now - I believe it's working as it should. In terms of your test data, did you take the hrrr and upscale it to match the mrms array size? I believe the data mask has to be the same dimensions and spacial resolution in order to properly work, correct? Thanks again, this is pretty awesome. I plan on spending more time this evening experimenting with it.

**Comment by tsupinie** [comment]
Yeah, that's correct. I used a nearest neighbor method on the HRRR CRAIN, CSNOW, CICEP, and CFRZR fields to get them to the MRMS grid and combined them onto a single grid. It's a super dumb method, and any method to do that operationally should probably be a little smarter. But I was going for test data, not an operational product, haha.

**Comment by tsupinie** [comment]
Now that v4 has been released with this feature, I'll go ahead and close this.
